### PR TITLE
Properly handle compiler versions

### PR DESCRIPTION
**Pull request by lukehoffmann**
This PR covers issue #318

`Compiler.get_version()` will return a tuple of integer values, e.g. `(8, 5, 0)` instead of the string `"8.5.0"`. This makes it possible to do greater-than and lesser-than comparisons with required compiler versions.

The different compiler classes will also verify that the version output contains the expected compiler name.

### CODE REVIEW

File: source/fab/tools/compiler.py [hunk 1]
```diff
@@ -92,68 +93,95 @@ def check_available(self) -> bool:
...
+            return True
+        except RuntimeError as err:
+            # Compiler does not exist, or version could not be handled:
+            self.logger.error(f'Error getting compiler version: {err}')
             return False
 
-        # An empty string is returned if some other error occurred when trying
-        # to get the compiler version.
-        return version != ""
-
```
Comment (allynt) [inline source/fab/tools/compiler.py:103]: I'm not very familiar w/ this code - is there a chance an exception other than `RuntimeError` could happen?  In that case should you remove the return statement in the try block and re-insert the `version != ""` here?

File: source/fab/tools/compiler.py [hunk 2]
```diff
@@ -92,68 +93,95 @@ def check_available(self) -> bool:
...
+            raise RuntimeError(f"Unexpected version output format for "
+                               f"compiler '{self.name}'. Should be numeric "
+                               f"<n.n[.n, ...]>: {version_string}") from err
+
+        # Expect at least 2 integer components, i.e. major.minor[.patch, ...]
+        if len(version) < 2:
+            raise RuntimeError(f"Unexpected version output format for "
+                               f"compiler '{self.name}'. Should have at least "
+                               f"two parts, <n.n[.n, ...]>: {version_string}")
+
```
Comment (allynt) [inline source/fab/tools/compiler.py:139]: Why is this changed from a log warning to a runtime error?

Comment (lukehoffmann) [reply source/fab/tools/compiler.py:139]: I didn't mention in the summary, but part of this change is the design decision that `get_version()` should only return a valid version tuple. if it can't determine a version, it raises a runtime error.
Previously it could return an empty string in some cases, but we thought that just made things unclear, and increased the amount of checking you might need to do when calling `get_version()`.

File: source/fab/tools/compiler.py (hunk 1 above)
Comment (lukehoffmann) [reply source/fab/tools/compiler.py:103]: `get_version()` will no longer return an empty string, and there is no other 'empty' type value to check for. It will always raise a `RuntimeError` if it can't determine the version.
`RuntimeError` is the only expected error type from the underlying `self.run()` function (which is in `tool.py`). It handles the `FileNotFoundError` if the compiler isn't found at all.

File: source/fab/tools/compiler.py (hunk 2 above)
Comment (hiker) [reply source/fab/tools/compiler.py:139]: Agreed - if we can't detect the version number (be it because the compiler has changed, or we have a bug in Fab), we need to abort. In LFRic we will have compiler flags added depending on compiler version, so it is important that we get the version correct.

**Review by allynt — APPROVED** [review]
All my comments have been addressed.  So I am approving this.
